### DISCUSSION TITLE
Show the latest vulnrebilityreports

### DISCUSSION THREAD

**Discussion by nissessenap** [aquasecurity/starboard#858] [Development]
Hey you all.
I'm a user of the giantswarm/starboard-exporter and it's really nice. But one of my main problems with it right now is that i see metrics for a bunch of CVE:s for old vulnrebilityreports in old replicasets.
I have created a issue with them how to solve it at the starboard-exporter side but we could defiantly do things to make that easier upstream.
If we could define the currently active replicates in the  vulnrebilityreports that would really be helpful.

There is an old case that I guess talk about a potential solution by using annotation on the latest vulnrebilityreports https://github.com/aquasecurity/starboard/issues/17
If that would be the case we could easily check for this in the exporter. I see that you have closed this issue.
Would you at all be willing to add this or similar functionality to this? 

Another solution could be to implement https://github.com/aquasecurity/starboard/discussions/668 and have a feature where only active replicasets gets vulnrebilityreports generated for them.

**Comment by itaysk** [comment]
If I understand correctly, there could be an old replicaset in transition, which you would still want to scan. So even if we did have #17 (a latest annotation) I don't think it's a good solution for this problem (since you would be missing real vulnerabilities), but nevertheless that specific issue about history is out of scope for Starboard.
I think #668 is a healthy suggestion regardless, and if it helps you than there's even more motivation do to it. @danielpacak WDYT?

**Comment by danielpacak** [comment]
Before I share my feedback lets come up with a minimal scenario to demonstrate the problem described by @NissesSenap and make sure we are talking about the same thing.

1. Start off by creating the nginx Deployment
   ```
   kubectl create deploy nginx --image nginx:1.16.0
   ```
2. Perform a few rolling updates of the nginx Deployment
   ```
   kubectl set image deploy/nginx nginx=nginx:1.16.1
   kubectl set image deploy/nginx nginx=nginx:1.17.0
   kubectl set image deploy/nginx nginx=nginx:1.17.1
   ```
3. Check that the controller-manager has created multiple ReplicaSets and only one is active (nginx:1.17.1)
   ```console
   $ kubectl get deploy -o wide
   NAME    READY   UP-TO-DATE   AVAILABLE   AGE   CONTAINERS   IMAGES         SELECTOR
   nginx   1/1     1            1           57m   nginx        nginx:1.17.1   app=nginx
   ```
   ```console
   $ kubectl get rs -o wide --sort-by .metadata.creationTimestamp
   NAME               DESIRED   CURRENT   READY   AGE   CONTAINERS   IMAGES         SELECTOR
   nginx-5dd4c884     0         0         0       58m   nginx        nginx:1.16.0   app=nginx,pod-template-hash=5dd4c884
   nginx-7f785d94f5   0         0         0       42m   nginx        nginx:1.16.1   app=nginx,pod-template-hash=7f785d94f5
   nginx-844f997cc9   0         0         0       41m   nginx        nginx:1.17.0   app=nginx,pod-template-hash=844f997cc9
   nginx-58b695f848   1         1         1       39m   nginx        nginx:1.17.1   app=nginx,pod-template-hash=58b695f848
   ```
   ```console
   $ kubectl tree deploy nginx
   NAMESPACE  NAME                            READY  REASON  AGE
   default    Deployment/nginx                -              58m
   default    ├─ReplicaSet/nginx-58b695f848   -              39m
   default    │ └─Pod/nginx-58b695f848-dvlqt  True           39m
   default    ├─ReplicaSet/nginx-5dd4c884     -              58m
   default    ├─ReplicaSet/nginx-7f785d94f5   -              42m
   default    └─ReplicaSet/nginx-844f997cc9   -              42m
   ```
4. If we deploy the Starboard Operator now it will scan the **active revision** as well as **all previous revisions** of the nginx Deployment. (Why? Because each ReplicaSet may point to a different container image reference.)

```console
$ kubectl tree deploy nginx
NAMESPACE  NAME                                                       READY  REASON  AGE
default    Deployment/nginx                                           -              86m
default    ├─ReplicaSet/nginx-58b695f848                              -              67m
default    │ ├─Pod/nginx-58b695f848-dvlqt                             True           67m
default    │ └─VulnerabilityReport/replicaset-nginx-58b695f848-nginx  -              2m35s
default    ├─ReplicaSet/nginx-5dd4c884                                -              86m
default    │ └─VulnerabilityReport/replicaset-nginx-5dd4c884-nginx    -              2m34s
default    ├─ReplicaSet/nginx-7f785d94f5                              -              70m
default    │ └─VulnerabilityReport/replicaset-nginx-7f785d94f5-nginx  -              2m34s
default    └─ReplicaSet/nginx-844f997cc9                              -              69m
default      └─VulnerabilityReport/replicaset-nginx-844f997cc9-nginx  -              2m35s
```

**Reply by nissessenap** [reply]
Yes just like this.

**Comment by danielpacak** [comment]
We backed off from #17 because we didn't want to store multiple instances of VulnerabilityReport for a given Kubernetes workload. The main reason for that was etcd as the default "storage driver" used by Starboard. etcd is not suitable for historical data or aggregates - it's supposed to store desired and current state of the cluster (and current vulnerabilities).

Beyond that, the `starboard.history.latest` label was redundant and somehow painful to manage. However, we can always figure out the active replica set by looking at the `deployment.kubernetes.io/revision=4` annotation that's used by controller-manager to find out the current revision to spin up or terminate pods. (We'll have to lookup the Deployment to compare the revisions, which are set on Deployment and ReplicaSet objects.)

So I think one option is to introduce a feature gate that only scans the active ReplicaSet and ignores previous revisions. If the feature gate is one you'd end up the VulnerabilityReport owned by the active ReplicaSet.

```console
$ kubectl tree deploy nginx
NAMESPACE  NAME                                                       READY  REASON  AGE
default    Deployment/nginx                                           -              139m
default    ├─ReplicaSet/nginx-58b695f848                              -              121m
default    │ ├─Pod/nginx-58b695f848-dvlqt                             True           121m
default    │ └─VulnerabilityReport/replicaset-nginx-58b695f848-nginx  -              56m
default    ├─ReplicaSet/nginx-5dd4c884                                -              139m
default    ├─ReplicaSet/nginx-7f785d94f5                              -              124m
default    └─ReplicaSet/nginx-844f997cc9
```

Notice that we do reconcile ReplicaSets and VulnerabilityReports in the same loop. For example, we rescan if you delete a VulnerabilityReport. We must check that it works as expected when we rollback an update.

BTW, a similar problem described by @NissesSenap with metrics we have when we generate an HTML report with Starboard CLI for a given namespace. It's listing all VulnerabilityReport instances in the given namespace and shows excessive vulnerability summaries. Whereas, it could also filter out by the `deployment.kubernetes.io/revision` annotation.

**Reply by itaysk** [reply]
correct me if I'm wrong but it could be that a previous replicaset is still active (meaning it has deployed pods) and is using an older image. In this case it's important to scan the older replicasets as well isn't it? if we ignore all empty replicasets instead of all previous replicasets it seems like the best balance between too much and too little information

**Reply by stone-z** [reply]
> it could be that a previous replicaset is still active (meaning it has deployed pods) and is using an older image... if we ignore all empty replicasets instead of all previous replicasets it seems like the best balance between too much and too little information

This is what I was thinking as well. I don't mind so much which the "latest" version is, I'm most interested in what's actually there. So having VulnerabilityReports for ReplicaSets with e.g. (Current > 0) OR (Desired > 0) would be more relevant than the latest revision.

**Reply by danielpacak** [reply]
You are right. Depending on the RollingUpgrade strategy set there will be a time window where pods are controlled by current and previous ReplicaSet. However, Starboard is not a runtime system. Eventually all pods will be controlled by the active ReplicaSet and eventually we want to show vulnerabilities in the running pods.

I suggested looking at the deployment.kubernetes.io/revision annotation to avoid all corner cases and rescanning scenarios when someone or K8s autoscaler will scale up or scale down based on traffic or some other criteria. Also, we plan to scan by Pods not by ReplicaSets one day to detect so called mutant containers.

**Reply by danielpacak** [reply]
So I think if we want to show vulnerabilities when a Deployment is being rolling updated (old and new pods running) then we should be looking at replicas count. If however we just care about the final state the deployment.kubernetes.io/revision annotations is maybe more appropriate. I don't have any strong opinion about that so we can vote :)

**Reply by itaysk** [reply]
I was thinking that a deployment could be stuck in flight for intentional or accidental reasons but I think you are right in that a deployment's purpose is to stay with one active replicaset and other cases are temporal or erroneous. So I agree that looking at latest revision is the correct solution for deployment.

Having said that, I still think that ignoring replicasets is a good idea, no? also it might solve the issue for other pod management controllers other than deployment (where the scenario I described above is legitimate, for example an a/b testing controller)

If we are planning to switch to scanning pods instead of replicasets then this would also be a good solution I think

**Reply by danielpacak** [reply]
That's a good point. We should probably also consider custom replication controllers and which strategy makes more sense in this case. So far I only thought about the default replicasetcontroller.

**Reply by nissessenap** [reply]
So I have played with this a bit locally and managed to get it to work for deployments and replicasets using `deployment.kubernetes.io/revision`.
I haven't tested for replicationController (I have never created one earlier so I will see how that works and if it got some similar annotation).
I haven't taken custom replication controllers in to consideration since it seems a bit out of scope for this discussion/issue. 

I will create a PR later today/tomorrow and you can look through it and see what you think.

**Comment by nissessenap** [comment]
I have created PR: https://github.com/aquasecurity/starboard/pull/870 and hopefully you think this solution is okay  from what we have discussed here.
ReplicationController don't even trigger a new creation of pods when you edit a existing one so I just ignored it, hopefully there aren't many people that still use that feature.
